src/client/ua_client_connect.c

```diff
@@ -1291,12 +1317,40 @@ __Client_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
     UA_UNLOCK(&client->clientMutex);
 }
 
+static void closeListeningReverseConnectSocket(UA_Client *client) {
+    UA_LOCK(&client->clientMutex);
+    if (client->channel.state != UA_SECURECHANNELSTATE_REVERSE_LISTENING) {
+        UA_UNLOCK(&client->clientMutex);
+        return;
+    }
+
+    client->channel.listenerDisconnected = true;
+    client->channel.connectionManager->closeConnection(client->channel.connectionManager, client->channel.connectionId);
+
+    UA_EventLoop *el = client->config.eventLoop;
```

Comment:
The cleanup of the listen socket can be done asynchronous in the next EventLoop iteration.
No need to run the EventLoop here.

Just reset `client->channel.connectionId` (to zero or to the id of the new "active socket").
Then, at some later point you will drop into the networkCallback with a closing socket with `connectionId != client->channel.connectionId`.
That tells you that this is the cleanup for an old connection and there is nothing left to do in the networkCallback (no SecureChannel to close).

Reply:
The suggested change makes the state callback miss the return to UA_SECURECHANNELSTATE_CLOSED if UA_Client_disconnect() is called while in state UA_SECURECHANNELSTATE_REVERSE_LISTENING. I'd like to keep the event loop run calls here if that's fine with you.

Reply:
Done